crypto/kdf/tls1_prf.c

```diff
@@ -31,145 +35,134 @@ typedef struct {
     /* Buffer of concatenated seed data */
     unsigned char seed[TLS1_PRF_MAXBUF];
     size_t seedlen;
-} TLS1_PRF_PKEY_CTX;
+};
 
-static int pkey_tls1_prf_init(EVP_PKEY_CTX *ctx)
+static EVP_KDF_IMPL *kdf_tls1_prf_new(void)
 {
-    TLS1_PRF_PKEY_CTX *kctx;
+    EVP_KDF_IMPL *impl;
 
-    if ((kctx = OPENSSL_zalloc(sizeof(*kctx))) == NULL) {
-        KDFerr(KDF_F_PKEY_TLS1_PRF_INIT, ERR_R_MALLOC_FAILURE);
-        return 0;
-    }
-    ctx->data = kctx;
+    if ((impl = OPENSSL_zalloc(sizeof(*impl))) == NULL)
+        KDFerr(KDF_F_KDF_TLS1_PRF_NEW, ERR_R_MALLOC_FAILURE);
+    return impl;
+}
 
-    return 1;
+static void kdf_tls1_prf_free(EVP_KDF_IMPL *impl)
+{
+    kdf_tls1_prf_reset(impl);
+    OPENSSL_free(impl);
 }
 
-static void pkey_tls1_prf_cleanup(EVP_PKEY_CTX *ctx)
+static void kdf_tls1_prf_reset(EVP_KDF_IMPL *impl)
 {
-    TLS1_PRF_PKEY_CTX *kctx = ctx->data;
-    OPENSSL_clear_free(kctx->sec, kctx->seclen);
-    OPENSSL_cleanse(kctx->seed, kctx->seedlen);
-    OPENSSL_free(kctx);
+    OPENSSL_clear_free(impl->sec, impl->seclen);
+    OPENSSL_cleanse(impl->seed, impl->seedlen);
+    memset(impl, 0, sizeof(*impl));
 }
 
-static int pkey_tls1_prf_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
+static int kdf_tls1_prf_ctrl(EVP_KDF_IMPL *impl, int cmd, va_list args)
 {
-    TLS1_PRF_PKEY_CTX *kctx = ctx->data;
-    switch (type) {
-    case EVP_PKEY_CTRL_TLS_MD:
-        kctx->md = p2;
-        return 1;
+    const unsigned char *p;
+    size_t len;
+    const EVP_MD *md;
 
-    case EVP_PKEY_CTRL_TLS_SECRET:
-        if (p1 < 0)
+    switch (cmd) {
+    case EVP_KDF_CTRL_SET_MD:
+        md = va_arg(args, const EVP_MD *);
+        if (md == NULL)
             return 0;
-        if (kctx->sec != NULL)
-            OPENSSL_clear_free(kctx->sec, kctx->seclen);
-        OPENSSL_cleanse(kctx->seed, kctx->seedlen);
-        kctx->seedlen = 0;
-        kctx->sec = OPENSSL_memdup(p2, p1);
-        if (kctx->sec == NULL)
+
+        impl->md = md;
+        return 1;
+
+    case EVP_KDF_CTRL_SET_TLS_SECRET:
+        p = va_arg(args, const unsigned char *);
+        len = va_arg(args, size_t);
+        OPENSSL_clear_free(impl->sec, impl->seclen);
+        OPENSSL_cleanse(impl->seed, impl->seedlen);
+        impl->seedlen = 0;
```

Comment:
That behavior is inherited from `EVP_PKEY_CTX_set_tls1_prf_secret()`, is documented (see: doc/man3/EVP_PKEY_CTX_set_tls1_prf_md.pod for EVP_PKEY and doc/man7/EVP_KDF_TLS1_PRF.pod for EVP_KDF) and appears to be intentional, possibly so that the context seed buffer could be reset without having to call `EVP_PKEY_derive_init()` again.

Remember that `EVP_PKEY_CTX_add1_tls1_prf_seed()` (as well as `EVP_KDF_ctrl(kctx, EVP_KDF_CTRL_ADD_TLS_SEED, ...)`) appends to whatever is in the context seed buffer rather than replacing it so there has to be some way to reset it to an empty buffer for the case where the context is to be reused. I don't know the history behind it, but resetting the context seed buffer when setting the secret value is set is the solution that was chosen. (Therefore you have to set the secret value before the first chunk of the seed value.) Of course we now have a `EVP_KDF_reset()` function that will reset everything.

I know you want a clean abstraction in luaossl and probably have crypto-agility in mind. The EVP_KDF API supports several KDF and PRF algorithms that are not equivalent or interchangeable; some, such as the TLS PRF, are special purpose or only used for a particular protocol so there is no need for them all to work exactly the same as each other.

I will have a look at the code in libssl where this algorithm is used and think about what you have suggested.

@levitte what do you think? Is this something that we should change?

Reply:
When I designed `EVP_MAC`, I attempted to do so with a clear mind of what it should do and not let the `EVP_PKEY` mac implementations influence me too much.  In essence, the `EVP_PKEY` implementations had to adapt to the `EVP_MAC` API, not the other way around.

So I guess that's my best advice, to try and make `EVP_KDF` into something that makes sense in itself.  If you feel that it doesn't really make sense for it to reset the seed here, then don't!  Finding it hard to justify this behavior is an indicator that you might want to change that and then have the `EVP_PKEY` implementation do an extra dance, so to say.  `EVP_KDF_reset` might be useful, depending on how much it resets.

Throwing in another idea, you might want to consider extra controls, such as `EVP_KDF_CTRL_RESET_TLS_SEED`, which the `EVP_PKEY` implementation can then call appropriately.

Reply:
Now that I think of it, I would add a RESET control for any control that *adds* information, just for completeness and to give the users a sense of control.

Reply:
> Now that I think of it, I would add a RESET control for any control that _adds_ information, just for completeness and to give the users a sense of control.

Good idea. I will add RESET controls to go along with EVP_KDF_CTRL_ADD_TLS_SEED and EVP_KDF_CTRL_ADD_HKDF_INFO.

Reply:
I have made the change to add RESET controls; `EVP_KDF_ctrl(ctx, EVP_KDF_CTRL_SET_TLS_SECRET, ...)` no longer resets the context seed buffer so the commit fixes #7728 .